Store and sync PostCommandLogs with Swarm

agazso opened the pull request:
Now we can store all post data in `PostCommandLog` locally and (in theory, hasn't yet well tested) then they can be synced automatically with other `PostCommandLog`s. The `PostCommandLog`s can be stored anywhere, for example on Swarm. This means it also becomes possible to sync between multiple devices.

There are helper functions to achieve this (e.g.
```ts
syncPostCommandLogWithSwarm: (postCommandLog: PostCommandLog, swarmApi: Swarm.Api): Promise<PostCommandLog>
```

Current issues:
- Right now I kept the implementation the most straightforward so it's not too efficient. For example right now there is no caching or logic to find quickly the last synced version, so we download the whole log and merge it with the whole log. The plan is to write tests for functionality first, then we can start thinking about optimizations.
- The `PostCommandLog` type now only consists a simple `PostCommand` array, so sometimes it feels extra work to pass around logs. On the other hand I can clearly see that in the future we can put extra data there, especially if we do conversions for other platforms and standards (e.g. ActivityStreams).

The next step are:
- [x] to make the Swarm API cleaner and more parametrisable (e.g. gatewayAddress).
- [x] create a separate Swarm Social layer which handles `PostCommandLog`s to keep Social API pure
- [x] cleanup and document Social API
- [x] write tests for edge-cases

After that is finished
- [x] when sharing a post use the Social API to post on Swarm
- [x] keep a `PostCommandLog` per feed on the device

vujevits commented on src/swarm-social/swarmStorage.ts:
```diff
@@ -0,0 +1,377 @@
+import { RecentPostFeed } from '../social/api';
+import { syncPostCommandLogWithStorage } from '../social/sync';
+
+export interface SwarmPostOptions {
```
options is probably not a good name for this

vujevits commented on src/swarm-social/swarmStorage.ts:
```diff
@@ -0,0 +1,377 @@
+
+export interface SwarmPostOptions {
+    imageResizer: (image: ImageData, path: string) => Promise<string>;
+    modelHelper: ModelHelper;
```
I would only pass here the needed function(s)

agazso commented:
The next step from here:
- [ ] Add resharing functionality. This can be done on Post or PostCommand level. We may need to add extra properties to support that functionality.

Todos that needs to be fixed later:

- [ ] Syncing works only when assuming there is only one device doing the updates. In order to make it work with multiple devices we need to add to StorageSyncUpdate:
  - [ ] the list of posts that were not uploaded
  - [ ] the list of posts that were deleted
  - [ ] the device identifier as source
- [ ] If sync fails it throws an exception, but it's possible that a partial update was done. In order to handle failed uploads, we need the list above and to make sure the exceptions are caught during sync and handled properly.
- [ ] if the local PostCommandLog is updated (for example a new photo is shared during syncing) we try to do a sync as soon as possible and that can lead to a sync loop, especially when combined with the above issue. The solution is to implement a smarter retry mechanism.
- [ ] The current solution works only assuming a single feed, because topics are not set correctly.

vujevits commented on src/actions/Actions.ts:
```diff
@@ -177,94 +189,102 @@ export const AsyncActions = {
+                        posts: [],
+                        authorImage: author.image,
+                        name: author.name,
+                        url: `bzz-feed:/?user=${address.user}`,
```
there should be a constant or helper function to assemble the url

agazso replied:
fixed

vujevits commented on src/actions/Actions.ts:
```diff
@@ -177,94 +189,102 @@ export const AsyncActions = {
 
-                    const author = getState().author;
-                    dispatch(Actions.updatePostIsUploading(post, true));
+                // Re-check if there were an update to the command log during syncing
```
will this still be an issue without gateway? (running swarm on mobile)

agazso replied:
Yes, it's still an issue, just the probability is smaller. Because we wait after every feed update  one second, and there are minimum 2 feed updates when syncing, it still allows enough time to accidentally run into this and the result is a post that is stuck in the spinning state.

vujevits commented on src/actions/Actions.ts:
```diff
@@ -283,26 +303,13 @@ export const AsyncActions = {
-    },
     restoreFromBackup: (backupText: string, secretHex: string): Thunk => {
-        return async (dispatch, getState) => {
+        return async (dispatch, getState: () => AppState) => {
```
why was that needed?

agazso replied:
fixed

vujevits commented on src/actions/Actions.ts:
```diff
@@ -332,15 +339,30 @@ const mergeImages = (localImages: ImageData[], uploadedImages: ImageData[]): Ima
     return allPosts;
 };
+
+const getSwarmStorageSyncer = (signFeedDigest: Swarm.FeedDigestSigner, feed: LocalFeed) => {
```
It might be better to have this in a separate file

agazso replied:
I was thinking the same, but then I couldn't decide where would I put it. I wouldn't put it under `swarm` or `swarm-social` because I think of them as separate generic packages and this code is specific to the app.

vujevits commented on src/reducers/index.ts:
```diff
@@ -433,7 +441,7 @@ const appStateReducer = (state: AppState = defaultState, action: Actions): AppSt
 class FelfelePersistConfig implements PersistConfig {
     public transforms = [immutableTransformHack({
-        whitelist: ['contentFilters', 'feeds', 'ownFeeds', 'rssPosts', 'localPosts', 'postUploadQueue'],
+        whitelist: ['contentFilters', 'feeds', 'ownFeeds', 'rssPosts', 'localPosts', 'postUploadQueue', 'postCommandLog'],
```
We shouldn't need the immutableTransformHack on postCommandLog

agazso replied:
fixed

vujevits commented on src/social/api.ts:
```diff
@@ -1,124 +1,222 @@
+     * If the command is stored on Swarm, this is set to the {@link Swarm.Epoch}
+     * of the update
+     */
     epoch?: Swarm.Epoch;
```
just a note for the future: 
to generalize this, we could pick some other name instead of epoch/previousEpoch, and using a union type when we want to support other platforms with different concepts

agazso replied:
Yes, that was my idea too, just didn't wanted to generalise it now. Actually if we wouldn't use epochs just (bzz) links, it would be much flexible and less tied to Swarm. On the other hand we would lose the ability to sort by epoch, then we would have to provide a sorter function on the interface, so I just left it as is for now.

vujevits commented on src/social/api.ts:
```diff
@@ -1,124 +1,222 @@
+ * Shares a new post and puts it in the command log
+ *
+ * @param post the post to be shared
+ * @param source the unique identifier of a user/device
```
are there any privacy/security implications?

agazso replied:
Yes, with public posts it leaks the data what device you used to create the post. I was thinking of several ideas how this can be mitigated:
- If you know the devices ahead (because there is a pairing process for example) you can use some derivation from the device ids, that has some shared secret involved (e.g. hmac). In this case the device info is not publicly visible and if we add the timestamp to the value to be hashed it will change with every post. The downside is that to track the source we must calculate all the hash of the known device ids to identify the source.
- An alternative is to use the `privateKey` as a symmetric encryption key with the `timestamp` as nonce to encrypt the source. I read somewhere that this may not be a good idea, so further research is needed.
- Or actually we can just use random if we don't really care about which device created the data. The sorting/merging algorithm will then work correctly assuming there are no conflicting ids which may break the invariants, but if the random space is big then this is not an issue.

vujevits replied:
There are 2 things which come to my mind:
- it shouldn't be _the_ device id, or anything which is used for other then our application, because that would definitely mean a privacy leak
- maybe (after integration swarm light-node) we can use something specific to the node. That would satisfy the above requirement, and it's also different from device to device for the same user

vujevits commented on src/social/api.ts:
```diff
@@ -251,237 +312,287 @@ export const removePost = (
+ *
+ * @returns the epoch or undefined if the log is empty
+ */
+export const getPreviousCommandEpochFromLog = (postCommandLog: PostCommandLog): Swarm.Epoch | undefined => {
```
[Don't use undefined as a means of denoting validity](https://basarat.gitbooks.io/typescript/docs/javascript/null-undefined.html)
I agree with this.

agazso replied:
I use this pattern at several places and at some places we use nulls as well, so let's discuss this further in the code conventions discussion #150 .

Also in the `PostCommand` when `epoch` is `undefined` it means that it hasn't been uploaded and this is consistent throughout the API. I am open to suggestions to improve this.

vujevits commented on src/swarm/Swarm.ts:
```diff
@@ -1,19 +1,19 @@
 import { Buffer } from 'buffer';
 
-export const DefaultGateway = 'https://swarm.felfele.com';
+export const DefaultGateway = 'http://localhost:8500';
```
is this for testing?

agazso replied:
fixed

vujevits commented on src/swarm/Swarm.ts:
```diff
@@ -137,23 +137,44 @@ interface FeedTemplate {
+
+const downloadUserFeedTemplate = async (swarmGateway: string, address: FeedAddress): Promise<FeedTemplate> => {
+    const addressPart = calculateFeedAddressQueryString(address);
+    const response = await downloadFeed(swarmGateway, `bzz-feed:/?${addressPart}&meta=1`);
```
as mentioned before, we interpolate / concat lot of swarm related urls, these should be handled at one place

agazso replied:
I agree, but I didn't want to spend too much time on the Swarm library, because I am already looking for a replacement (e.g. Erebos), because maintaining this is just too much effort.

vujevits commented on src/swarm/Swarm.ts:
```diff
@@ -166,70 +187,107 @@ export const updateUserFeed = async (feedTemplate: FeedTemplate, identity: Priva
+    readonly swarmGateway: string;
+}
+
+export interface WriteableApi extends BaseApi {
```
Reader / Writer might be better.

agazso replied:
I don't like Readable / Writeable either, right now my best idea is `DownloadApi`, `UploadApi`, `FeedDownloadApi` with `FeedUpdateApi`)

vujevits replied:
I'm good with Download(er) too.

vujevits commented:
👍 
well documentation too!

agazso commented:
As far as I'm concerned we can merge it now to master. There are known issues:
- [ ] Resyncing can get stuck in a loop, now it waits 30 seconds before a resync to mitigate it somewhat
- [ ] I changed the deletion to update the PostCommandLog and sync, but this has not been tested yet

I added a screen to change the swarm gateway address, that closed #134.

vujevits commented:
I tried it, works, but there was a SyncPostCommandLog error quite a few times when I shared a post, BUT it still worked, managed to follow from another device. ATM cannot reproduce it, so let's merge.
